File: app/src/main/java/ska/sudoku/SudokuHelper.java
package ska.sudoku;

import android.graphics.Point;

import java.util.List;

class SudokuHelper {

    static Point getSection(final int cell, final int max) {
        Point p = new Point();
        int itemsPerSection = max / 3;

        p.x = (cell % max) / itemsPerSection;
        p.y = cell / (max * itemsPerSection);

        return p;
    }

    static boolean checkValue(final List<Cell> grid, final Cell cell, final int value, final int maxValue) {
        return checkRow(grid, cell, value, maxValue) && checkColumn(grid, cell, value, maxValue) && checkSection(grid, cell, value, maxValue);
    }

    private static boolean checkRow(final List<Cell> grid, final Cell cell, final int value, final int maxValue) {
        int row = cell.getNr() / maxValue;
        for (int i = 0; i < maxValue; i++) {
            if (grid.get(i + maxValue * row).getValue() == value)
                return false;
        }
        return true;
    }

    private static boolean checkColumn(final List<Cell> grid, final Cell cell, final int value, final int maxValue) {
        int column = cell.getNr() % maxValue;
        for (int i = 0; i < maxValue; i++) {
            if (grid.get(i * maxValue + column).getValue() == value)
                return false;
        }
        return true;
    }

    private static boolean checkSection(final List<Cell> grid, final Cell cell, final int value, final int maxValue) {
        Point p = SudokuHelper.getSection(cell.getNr(), maxValue);
        int itemsPerSection = maxValue / 3;
        int sectionRow = p.y;
        int sectionColumn = p.x;

        for (int i = 0; i < itemsPerSection; i++) {
            int firstRow = i + sectionRow * maxValue * 3 + sectionColumn * itemsPerSection;
            for (int j = 0; j < itemsPerSection; j++) {
                int currentCell = firstRow + j * maxValue;
                if (cell.getNr() != currentCell)
                    if (grid.get(currentCell).getValue() == value)
                        return false;
            }
        }
        return true;
    }
}

File: app/src/main/java/ska/sudoku/SudokuHelper.kt
package ska.sudoku

fun isEvenSection(cell: Int, maxValue: Int): Boolean {
    val (x, y) = getSection(cell, maxValue)
    return (x + y) % 2 == 0
}

fun getSection(cell: Int, max: Int): Pair<Int, Int> {
    val itemsPerSection = max / 3

    val x = cell % max / itemsPerSection
    val y = cell / (max * itemsPerSection)

    return Pair(x, y)
}

fun checkValue(grid: List<Cell>, cell: Int, value: Int, maxValue: Int): Boolean {
    return checkRow(grid, cell, value, maxValue) &&
            checkColumn(grid, cell, value, maxValue) &&
            checkSection(grid, cell, value, maxValue)
}

private fun checkRow(grid: List<Cell>, cell: Int, value: Int, maxValue: Int): Boolean {
    val row = cell / maxValue
    return (0 until maxValue)
            .none { value == grid[it + maxValue * row].value }
}

private fun checkColumn(grid: List<Cell>, cell: Int, value: Int, maxValue: Int): Boolean {
    val column = cell % maxValue
    return (0 until maxValue)
            .none { value == grid[it * maxValue + column].value }
}

private fun checkSection(grid: List<Cell>, cell: Int, value: Int, maxValue: Int): Boolean {
    val (sectionColumn, sectionRow) = getSection(cell, maxValue)
    val itemsPerSection = maxValue / 3

    return (0 until itemsPerSection)
            .map { it + sectionRow * maxValue * 3 + sectionColumn * itemsPerSection }
            .none { firstRow ->
                (0 until itemsPerSection)
                        .map { firstRow + it * maxValue }
                        .filter { cell != it }
                        .any { value == grid[it].value }
            }
}
